### PR TITLE
commit

### PR DESCRIPTION
**Pull request by Muhammad-Miqdad-Ahmad**
My react project

### CODE REVIEW

File: back-end/Controllers/userController.js [hunk 1]
```diff
@@ -0,0 +1,137 @@
+const userData = require("../Schema/userData");
+const userDetails = require("../Schema/Details");
+const jwt = require('jsonwebtoken');
+const bcrypt = require('bcrypt');
+const SALT = 10
```
Comment (Azid-ops) [inline back-end/Controllers/userController.js]: Place all the Variables which are related to functionality inside the respective controllers or middlewares

File: back-end/Controllers/userController.js [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+const Registration = async (req, res) => {
+    console.log("starting")
```
Comment (Azid-ops) [inline back-end/Controllers/userController.js]: Before Pushing your code remove all the console

File: back-end/server.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
+const express = require('express')
+const UserRouter = require('./routes/UserRoutes');
+const { default: mongoose } = require('mongoose');
```
Comment (Azid-ops) [inline back-end/server.js]: what you used default:mongoose instead of only mongoose

File: front-end/src/App.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+// import Home from './Components/Home';
+import Login from './Components/Login';
+import Registration from './Components/Registration';
+import { Route, createBrowserRouter, createRoutesFromElements, RouterProvider, Outlet } from "react-router-dom";
+import Update from './Components/Update';
+import HomePage from './Components/HomePage';
+
+
+function App() {
+  const router = createBrowserRouter(createRoutesFromElements(
```
Comment (Azid-ops) [inline front-end/src/App.js]: This is correct but you should write JS Preferred code
```
<Router>
    <Routes>
        // Your Routes
        <Route />
     </Routes>
</Router>
```

File: front-end/src/App.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
+import './App.css';
+import DataEntry from './Components/DataEntry';
+// import Home from './Components/Home';
+import Login from './Components/Login';
+import Registration from './Components/Registration';
+import { Route, createBrowserRouter, createRoutesFromElements, RouterProvider, Outlet } from "react-router-dom";
```
Comment (Azid-ops) [inline front-end/src/App.js]: import like this
import { BrowserRouter as Router, Routes,Route} from 'react-router-dom";

this way we wont need any other things like createBrowserRouter and others

File: front-end/src/App.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+      <Route path='/Registration' element={<Registration />} />
+      <Route path='/Home' element={<HomePage />} />
+      <Route path='/Details' element={<DataEntry />} />
+      <Route path='/Update' element={<Update />} />
+    </Route>
+  ))
+
+  return (
+    <div className="App">
+      <RouterProvider router={router} />
```
Comment (Azid-ops) [inline front-end/src/App.js]: You should write Routers code directly in html , no need to create a variable for that, we dont want compiler to do extra stuff

File: front-end/src/Components/DataEntry.js [hunk 7]
```diff
@@ -0,0 +1,35 @@
+import React, { useState } from 'react'
+import { useNavigate } from 'react-router-dom';
+import { onChangeHandler } from './Fucntions';
+import { onClickHandler } from './Fucntions';
```
Comment (Azid-ops) [inline front-end/src/Components/DataEntry.js]: Use Appropriate Names like Functions and always use goodnames

File: front-end/src/Components/DataEntry.js [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+export default function DataEntry() {
+    const [data, setData] = useState('');
+    const navigate = useNavigate();
+
+    return (
+        <div className='text-center'>
+            <form onSubmit={e => {
+                console.log("onc click was called in the DataEntry.js");
```
Comment (Azid-ops) [inline front-end/src/Components/DataEntry.js]: Take extra care of indentation, clean code is always a good code

File: front-end/src/Components/DataEntry.js [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+                e.preventDefault();
+                onClickHandler('details', 'POST', navigate, data);
+            }}>
+                <label htmlFor="FirstName">FirstName:</label><br />
+                <input className='border-2 rounded-lg' type="text" id="FirstName" name="FirstName" onChange={event => onChangeHandler(event, data, setData)} /><br />
+                <label htmlFor="LastName">LastName:</label><br />
+                <input className='border-2 rounded-lg' type="text" id="LastName" name="LastName" onChange={event => onChangeHandler(event, data, setData)} /><br />
+                <label htmlFor="Age">Age:</label><br />
+                <input className='border-2 rounded-lg' type="number" id="Age" name="Age" onChange={event => onChangeHandler(event, data, setData)} /><br />
```
Comment (Azid-ops) [inline front-end/src/Components/DataEntry.js]: Again Clean Code Issue this should be like this
```
<input className='border-2 rounded-lg' 
    type="number" 
    id="Age" 
    name="Age"  
    onChange={event => onChangeHandler(event, data, setData)}
 />
<br />
```

File: front-end/src/Components/DataEntry.js [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+                <p htmlFor="Gender">Gender:</p>
+                <label htmlFor="Gender">Male: </label>
+                <input className='border-2 rounded-lg' type="radio" id="Male" name="Gender" onChange={event => onChangeHandler(event, data, setData)} />&nbsp;&nbsp;
+                <label htmlFor="Gender">Female: </label>
+                <input className='border-2 rounded-lg' type="radio" id="Female" name="Gender" onChange={event => onChangeHandler(event, data, setData)} /><br />
+                <label htmlFor="Profession">Profession:</label><br />
+                <input className='border-2 rounded-lg' type="Profession" id="Profession" name="Profession" onChange={event => onChangeHandler(event, data, setData)} /><br />
```
Comment (Azid-ops) [inline front-end/src/Components/DataEntry.js]: Id should not be same as name use a diffferent word for id

File: front-end/src/Components/Fucntions.js [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+    const temp = {};
+    temp[event.target.name] = event.target.value;
+
+    if (event.target.name === "Gender")
+        temp[event.target.name] = event.target.id;
+
+    if (data)
+        setData({ ...data, ...temp });
+    else
+        setData(temp);
```
Comment (Azid-ops) [inline front-end/src/Components/Fucntions.js:11]: This is good code, very good

File: front-end/src/Components/Fucntions.js [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+export const onClickHandler = async (route, method, navigate, data) => {
+    console.log("in the on click handler for the "+method);
```
Comment (Azid-ops) [inline front-end/src/Components/Fucntions.js]: remove console

File: front-end/src/Components/Fucntions.js [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+    const response = await fetch(`/${route}`, {
+        method: method,
```
Comment (Azid-ops) [inline front-end/src/Components/Fucntions.js:16]: what is this method is this a variable?
if this is a varibale then you don't need to write it in a variable unless it is not using in more than one place you can simply write POST or GET etc

File: front-end/src/Components/Fucntions.js [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+        console.log("navigate");
+        navigate('/Home');
+    }
+}
+
+export async function apiCall(navigate, setError, data) {
+    if (data['userName'] !== "" && data['userPassword'] !== "" && data['userEmail'] !== "") {
+
+        const response = await fetch("/registration", {
+            method: "POST",
```
Comment (Azid-ops) [inline front-end/src/Components/Fucntions.js:39]: this is good code

File: front-end/src/Components/Fucntions.js [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+        }
+        else {
+            setError(true);
+        }
+    }
+}
+
+export const buttonHandlerSignOut = (navigate) => {
+    console.log("buttonHandlerSignOut was called");
+    localStorage.removeItem("userToken");
```
Comment (Azid-ops) [inline front-end/src/Components/Fucntions.js:58]: Very good that you used local Storage but kindly use local storage with extra care.

File: front-end/src/Components/Home.js [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+
+  const navigate = useNavigate();
+  return (
+    <div className='border-cyan-600 border-2'>
+      <h1>First Name: {props?.schemaFirstName}</h1>
+      <h1>Last Name: {props?.schemaLastName}</h1>
+      <p>Age: {props?.schemaAge}</p>
+      <p>Gender: {props?.schemaGender}</p>
+      <p>Profession: {props?.schemaProfession}</p>
+      <br />
```
Comment (Azid-ops) [inline front-end/src/Components/Home.js]: Try Avoiding BR, using one br is good or using two is good but using alot of br one after another is not good

File: front-end/src/Components/Login.js [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+                <input type='checkbox' className='form-check-input' onChange={() => { setPass(!pass) }} />
+                <label className="form-check-label" htmlFor='form-check-input'> Show Password</label>
+            </div>
+            <br />
+            {
+                error ? "Something Went Wrong!" : null
+            }
+            <br />
+            <div className='text-center'>
+                <Link to="/Registration">Don't have an account? click here to signup</Link>
```
Comment (Azid-ops) [inline front-end/src/Components/Login.js]: Don't use Link cuz it is deprecated it is old javascript try making your custom link

File: front-end/src/Components/Registration.js [hunk 18]
```diff
@@ -0,0 +1,80 @@
+import React, { useEffect } from 'react'
+import { useState } from 'react';
+import { Link, useNavigate } from 'react-router-dom';
+import { apiCall } from './Fucntions';
```
Comment (Azid-ops) [inline front-end/src/Components/Registration.js]: Try to give some spaces between different types of imports like

```
State1
State2

Page1
Page2
```

etc etc

**Review by Azid-ops — CHANGES_REQUESTED** [review]
Overall Code is very good, but i have requested some changes.

File: back-end/server.js (hunk 3 above)
Comment (Muhammad-Miqdad-Ahmad) [reply back-end/server.js]: Vs code did it on its own, I did not even know it was there. I apologise

File: front-end/src/App.js (hunk 4 above)
Comment (Muhammad-Miqdad-Ahmad) [reply front-end/src/App.js]: Okay I will look into it and change this

File: front-end/src/Components/Fucntions.js (hunk 13 above)
Comment (Muhammad-Miqdad-Ahmad) [reply front-end/src/Components/Fucntions.js:16]: okay I will see to that

File: front-end/src/Components/Fucntions.js (hunk 15 above)
Comment (Muhammad-Miqdad-Ahmad) [reply front-end/src/Components/Fucntions.js:58]: okay I will take care in the future

File: front-end/src/Components/Home.js (hunk 16 above)
Comment (Muhammad-Miqdad-Ahmad) [reply front-end/src/Components/Home.js]: okay I will take care. I only wrote this for development and just did not remove these. I will take care that this never happens. My major focus was on the backend

File: front-end/src/Components/Registration.js (hunk 18 above)
Comment (Muhammad-Miqdad-Ahmad) [reply front-end/src/Components/Registration.js]: okay I will do so

File: front-end/src/Components/Login.js (hunk 17 above)
Comment (Muhammad-Miqdad-Ahmad) [reply front-end/src/Components/Login.js]: Done

File: front-end/src/Components/DataEntry.js (hunk 9 above)
Comment (Muhammad-Miqdad-Ahmad) [reply front-end/src/Components/DataEntry.js]: completed

**Review by Azid-ops — APPROVED** [review]
Good Work!!!

**Pull request merged**